client/session/grpctransport/grpc.go

```diff
@@ -0,0 +1,172 @@
+	matches := getMethodRe().FindAllStringSubmatch(in, -1)
+	if len(matches) != 1 || len(matches[0]) != 3 {
+		return "", "", errors.Errorf("invalid method name %s", in)
+	}
+	return matches[0][1], matches[0][2], nil
+}
+
+func (gt *grpcTransportFactory) NewHandler(ctx context.Context, conn net.Conn) (session.TransportHandler, error) {
+	return newHandler(ctx, conn)
+}
+func (gt *grpcTransportFactory) NewCaller(ctx context.Context, conn net.Conn) (session.TransportCaller, error) {
+	return newCaller(ctx, conn)
+}
+
+type callbackSelector struct {
+	id     string
+	method string
+}
+
+type grpcHandler struct {
+	conn     net.Conn
+	handlers map[callbackSelector]session.HandleFunc
+}
+
+func newHandler(ctx context.Context, conn net.Conn) (*grpcHandler, error) {
+	gh := &grpcHandler{
+		conn:     conn,
+		handlers: make(map[callbackSelector]session.HandleFunc),
+	}
+	return gh, nil
+}
+
+func (gh *grpcHandler) Register(id, method string, f session.HandleFunc) error {
+	// type HandleFunc func(ctx context.Context, opts map[string][]string, stream Stream) error
+	gh.handlers[callbackSelector{id: id, method: method}] = f
+	return nil
+}
+
+func (gh *grpcHandler) Serve(ctx context.Context) error {
+	gs := grpc.NewServer()
+	go func() {
+		<-ctx.Done()
+		gh.conn.Close()
+	}()
+
+	services := make(map[string][]callbackSelector)
+
+	for sel := range gh.handlers {
+		sname := serviceName(sel.id, sel.method)
+		services[sname] = append(services[sname], sel)
+	}
+
+	for name, selectors := range services {
+		var streams []grpc.StreamDesc
+		for _, sel := range selectors {
+			func(sel callbackSelector) {
+				streams = append(streams, grpc.StreamDesc{
+					StreamName: streamName(sel.id, sel.method),
+					Handler: func(srv interface{}, stream grpc.ServerStream) error {
+						logrus.Debugf("handling %v", sel)
+						var s session.Stream
+						ctx := stream.Context()
+						md, _ := metadata.FromContext(ctx)
+						s = stream
+						return gh.handlers[sel](ctx, md, &s)
+					},
+					ServerStreams: true,
+					ClientStreams: true,
+				})
+			}(sel)
+		}
+		gs.RegisterService(&grpc.ServiceDesc{
+			ServiceName: name,
+			HandlerType: (*interface{})(nil),
+			Methods:     []grpc.MethodDesc{},
+			Streams:     streams,
+		}, (*interface{})(nil))
+	}
+	logrus.Debugf("serving")
+	(&http2.Server{}).ServeConn(gh.conn, &http2.ServeConnOpts{Handler: gs})
+	return nil
+}
+
+type grpcCaller struct {
+	cc *grpc.ClientConn
+}
+
+func newCaller(ctx context.Context, conn net.Conn) (*grpcCaller, error) {
+	dialOpt := grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+		return conn, nil
+	})
+
+	cc, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithInsecure())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create grpc client")
+	}
+
+	gc := &grpcCaller{
+		cc: cc,
+	}
+
+	return gc, nil
+}
+
+func (gc *grpcCaller) Call(ctx context.Context, id, method string, opt map[string][]string) (session.Stream, error) {
```

Comment:
@stevvooe The abstraction is just so that you can write isolated features(currently rsync context sending, client-side token queries and auth requests, ssh signing) that can be exposed through remote API, without the features knowing how they would be accessed on the wire. The remote API starts by making a POST request with exposed method names and then hijacks the tcp connection for a transport. The abstraction was added because no transport method seemed clearly better than rest. Currently, there are no plans to ship support for multiple transport implementations.